src/plots/cartesian/autorange.js

```diff
@@ -0,0 +1,324 @@
+            else if(newRange[0] > 0) {
+                newRange = [0, newRange[0]];
+            }
+            else {
+                newRange = [0, 1];
+            }
+        }
+        else {
+            newRange = [newRange[0] - 1, newRange[0] + 1];
+            if(ax.rangemode === 'nonnegative') {
+                newRange[0] = Math.max(0, newRange[0]);
+            }
+        }
+    }
+
+    // maintain reversal
+    if(axReverse) newRange.reverse();
+
+    return Lib.simpleMap(newRange, ax.l2r || Number);
+}
+
+/*
+ * calculate the pixel padding for ax._min and ax._max entries with
+ * optional extrapad as 5% of the total axis length
+ */
+function makePadFn(ax) {
+    // 5% padding for points that specify extrapad: true
+    var extrappad = ax._length / 20;
+
+    // domain-constrained axes: base extrappad on the unconstrained
+    // domain so it's consistent as the domain changes
+    if((ax.constrain === 'domain') && ax._inputDomain) {
+        extrappad *= (ax._inputDomain[1] - ax._inputDomain[0]) /
+            (ax.domain[1] - ax.domain[0]);
+    }
+
+    return function getPad(pt) { return pt.pad + (pt.extrapad ? extrappad : 0); };
+}
+
+function doAutoRange(ax) {
+    ax.setScale();
+
+    // TODO do we really need this?
+    var hasDeps = (ax._min && ax._max && ax._min.length && ax._max.length);
+
+    if(ax.autorange && hasDeps) {
+        ax.range = getAutoRange(ax);
+
+        ax._r = ax.range.slice();
+        ax._rl = Lib.simpleMap(ax._r, ax.r2l);
+
+        // doAutoRange will get called on fullLayout,
+        // but we want to report its results back to layout
+
+        var axIn = ax._input;
+        axIn.range = ax.range.slice();
+        axIn.autorange = ax.autorange;
+    }
+}
+
+function needsAutorange(ax) {
+    return ax.autorange || !!(ax.rangeslider || {}).autorange;
+}
+
+// expand: if autoranging, include new data in the outer limits
+// for this axis
+// data is an array of numbers (ie already run through ax.d2c)
+// available options:
+//      vpad: (number or number array) pad values (data value +-vpad)
+//      ppad: (number or number array) pad pixels (pixel location +-ppad)
+//      ppadplus, ppadminus, vpadplus, vpadminus:
+//          separate padding for each side, overrides symmetric
+//      padded: (boolean) add 5% padding to both ends
+//          (unless one end is overridden by tozero)
+//      tozero: (boolean) make sure to include zero if axis is linear,
+//          and make it a tight bound if possible
+function expand(ax, data, options) {
+    if(!needsAutorange(ax) || !data) return;
+
+    if(!ax._min) ax._min = [];
+    if(!ax._max) ax._max = [];
+    if(!options) options = {};
+    if(!ax._m) ax.setScale();
+
+    var len = data.length;
+    var extrapad = options.padded || false;
+    var tozero = options.tozero && (ax.type === 'linear' || ax.type === '-');
+
+    var i, j, k, v, di, dmin, dmax, ppadiplus, ppadiminus, includeThis, vmin, vmax;
+
+    function getPad(item) {
```

Comment:
We could maybe rename this function. `getPad` might be confused with the output of `makePadFn`

Reply:
haha I noticed that too... I'll change it.

Reply:
renamed -> https://github.com/plotly/plotly.js/pull/2437/commits/9913afe4549580bb4b0271a5cf645633a8dfdab1